### PR TITLE
Bcsd fit and predict

### PR DESCRIPTION
**Pull request by norlandrhagen**
Updated fit_and_predict task. Split `get_experiment()` calls into two separate calls for predict and train time periods.

### CODE REVIEW

File: cmip6_downscaling/methods/bcsd/tasks.py [hunk 1]
```diff
@@ -87,9 +93,31 @@ def fit_and_predict(
         print(f'found existing target: {target}')
         return target
 
-    # TODO
+    # # TODO
+    if run_parameters.variable in ABSOLUTE_VARS:
+        bcsd_model = BcsdTemperature(return_anoms=False)
+    elif run_parameters.variable in RELATIVE_VARS:
+        bcsd_model = BcsdPrecipitation(return_anoms=False)
```
Comment (jhamman) [inline cmip6_downscaling/methods/bcsd/tasks.py:100]: raise `ValueError` in `else`

File: cmip6_downscaling/methods/bcsd/tasks.py [hunk 2]
```diff
@@ -87,9 +93,31 @@ def fit_and_predict(
...
+
+    pointwise_model = PointWiseDownscaler(model=bcsd_model, dim="time")
+
+    # coarse_obs_rechunked_validated_ds
+    coarse_obs_full_time_ds = xr.open_zarr(coarse_obs_full_time_path)
+    # gcm_train_subset_full_time_ds
+    experiment_train_full_time_ds = xr.open_zarr(experiment_train_full_time_path)
+    # gcm_predict_rechunked_ds
+    experiment_predict_full_time_ds = xr.open_zarr(experiment_predict_full_time_path)
```
Comment (jhamman) [inline cmip6_downscaling/methods/bcsd/tasks.py:109]: I don't think these comments add enough to motivate their use here.

```suggestion
    coarse_obs_full_time_ds = xr.open_zarr(coarse_obs_full_time_path)
    experiment_train_full_time_ds = xr.open_zarr(experiment_train_full_time_path)
    experiment_predict_full_time_ds = xr.open_zarr(experiment_predict_full_time_path)
```

File: cmip6_downscaling/methods/bcsd/tasks.py [hunk 3]
```diff
@@ -77,7 +81,9 @@ def calc_spacial_anomalies(
 
 @task
 def fit_and_predict(
-    experiment_full_time_path: UPath,
+    # inputs: gcm_train_subset_full_time_path, coarse_obs_rechunked_validated_path, gcm_predict_rechunked_path
```
Comment (jhamman) [inline cmip6_downscaling/methods/bcsd/tasks.py:84]:
```suggestion
```

File: cmip6_downscaling/methods/bcsd/flow.py [hunk 4]
```diff
@@ -44,20 +44,25 @@
 
     # input datasets
     obs_path = get_obs(run_parameters)
-    experiment_path = get_experiment(run_parameters)
+    experiment_train_path = get_experiment(run_parameters, time_subset='train_period')
+    experiment_predict_path = get_experiment(run_parameters, time_subset='predict_period')
 
-    coarse_obs_path = coarsen_obs(obs_path, experiment_path, run_parameters)
+    coarse_obs_path = coarsen_obs(obs_path, experiment_train_path, run_parameters)
```
Comment (jhamman) [inline cmip6_downscaling/methods/bcsd/flow.py:50]: this is turning into `regrid` in #120, right?

File: cmip6_downscaling/methods/common/tasks.py [hunk 5]
```diff
@@ -74,13 +74,43 @@ def get_experiment(run_parameters: RunParameters) -> UPath:
...
+        return target
 
-    subset.to_zarr(target, mode='w')
+    ds = load_cmip(
+        source_ids=run_parameters.model, return_type='xr', variable_ids=run_parameters.variable
+    ).pipe(lon_to_180)
+
+    # subset = subset_dataset(
+    #     ds, run_parameters.variable, run_parameters.train_period.time_slice, run_parameters.bbox
+    # )
```
Comment (jhamman) [inline cmip6_downscaling/methods/common/tasks.py:109]:
```suggestion
```

No longer needed, right?

Comment (jhamman) [reply cmip6_downscaling/methods/common/tasks.py:109]: or is it?

Comment (norlandrhagen) [reply cmip6_downscaling/methods/common/tasks.py:109]: Ah, I think I added some confusion in by leaving in a partially finished idea of `get_experiment_training`. I think the updated `get_experiment` PR should fix this. https://github.com/carbonplan/cmip6-downscaling/pull/125#discussion_r829553643

**Pull request merged**